### PR TITLE
Optimize the `RamInode`: `DirEntry` mainly and page cache

### PR DESCRIPTION
**Pull request by lucassong-mh**

### CODE REVIEW

File: kernel/src/fs/ramfs/fs.rs [hunk 1]
```diff
@@ -351,29 +356,46 @@ impl DirEntry {
...
         } else {
-            self.children
-                .idxes_and_items()
-                .find(|(_, (child, _))| child.as_str().unwrap() == name)
-                .map(|(idx, (_, inode))| (idx + 2, inode.clone()))
+            let idx = *self.idx_map.get(name.as_bytes())?;
+            let target_inode = self
+                .children
+                .get(idx)
+                .map(|(n, inode)| {
```
Comment (tatetian) [inline kernel/src/fs/ramfs/fs.rs:363]: `n` appears to be a number, but it is not. How about `name_cstr256`?

File: kernel/src/fs/ramfs/fs.rs [hunk 2]
```diff
@@ -351,29 +356,46 @@ impl DirEntry {
...
     ) -> Option<(CStr256, Arc<RamInode>)> {
-        assert!(idx >= 2);
-        self.children.put_at(idx - 2, new_entry)
+        assert!(idx >= NUM_SPECIAL_ENTRIES);
+        let name = new_entry.0;
+        let idx_children = idx - NUM_SPECIAL_ENTRIES;
+
+        let substitute = self.children.put_at(idx_children, new_entry)?;
+        self.idx_map.remove(&substitute.0);
+        self.idx_map.insert(name, idx_children);
```
Comment (tatetian) [inline kernel/src/fs/ramfs/fs.rs:397]: The call to `remove` is unnecessary.  The subsequent call to `insert` will replace and return any old entry if it exists.

File: kernel/src/fs/ramfs/fs.rs [hunk 3]
```diff
@@ -519,7 +541,11 @@ impl RamInode {
 impl PageCacheBackend for RamInode {
     fn read_page_async(&self, _idx: usize, frame: &Frame) -> Result<BioWaiter> {
         // Initially, any block/page in a RamFs inode contains all zeros
-        frame.writer().fill(0);
+        frame
+            .writer()
+            .to_fallible()
+            .fill_zeros(frame.size())
```
Comment (tatetian) [inline kernel/src/fs/ramfs/fs.rs:547]: This change has nothing to do with optimizing dir entries, so it should be done in a separate commit.

File: kernel/src/fs/ramfs/fs.rs (hunk 2 above)
Comment (lucassong-mh) [reply kernel/src/fs/ramfs/fs.rs]: I'm afraid not, the `idx_map` now caches mapping "name -> idx", the `substitute_entry()` is used in `rename()`. In this case the `idx` is the same but the name is not, the old name and the new name appear in different KVs in `idx_map`, so we need to remove the old one then insert the new one.

**Review by tatetian — APPROVED** [review]
LGTM. Thanks for the contribution!
